org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java

```diff
@@ -437,51 +439,60 @@ public void setMojoExecutionMapping(Map<MojoExecutionKey, List<IPluginExecutionM
-        calculateExecutionPlan(mavenProject, List.of(ProjectRegistryManager.LIFECYCLE_DEFAULT), false, monitor)
-            .getMojoExecutions());
-    executionPlans.put(ProjectRegistryManager.LIFECYCLE_SITE,
-        calculateExecutionPlan(mavenProject, List.of(ProjectRegistryManager.LIFECYCLE_SITE), false, monitor)
-            .getMojoExecutions());
-    return executionPlans;
+    Map<String, List<MojoExecution>> plans = new LinkedHashMap<>();
+    addCalculatedLifecyclePlan(project, ProjectRegistryManager.LIFECYCLE_CLEAN, plans, context, exHandler, monitor);
+    addCalculatedLifecyclePlan(project, ProjectRegistryManager.LIFECYCLE_DEFAULT, plans, context, exHandler, monitor);
+    addCalculatedLifecyclePlan(project, ProjectRegistryManager.LIFECYCLE_SITE, plans, context, exHandler, monitor);
+    return plans;
+  }
+
+  private static void addCalculatedLifecyclePlan(MavenProject project, String lifecycle,
+      Map<String, List<MojoExecution>> plans, IMavenExecutionContext context, Consumer<? super CoreException> exHandler,
+      IProgressMonitor monitor) {
+    MavenExecutionPlan plan = calculateExecutionPlan(project, List.of(lifecycle), false, context, exHandler, monitor);
+    plans.put(lifecycle, plan.getMojoExecutions());
   }
 
   @Override
   public MavenExecutionPlan calculateExecutionPlan(Collection<String> tasks, IProgressMonitor monitor) {
-    return calculateExecutionPlan(getMavenProject(), tasks, false, monitor);
+    return calculateExecutionPlan(tasks, monitor, false);
   }
 
   @Override
   public MavenExecutionPlan setupExecutionPlan(Collection<String> tasks, IProgressMonitor monitor) {
-    return calculateExecutionPlan(getMavenProject(), tasks, true, monitor);
+    return calculateExecutionPlan(tasks, monitor, true);
+  }
+
+  private MavenExecutionPlan calculateExecutionPlan(Collection<String> tasks, IProgressMonitor monitor, boolean setup) {
+    MavenProject project = getMavenProject();
+    IMavenExecutionContext context = createExecutionContext();
+    return calculateExecutionPlan(project, tasks, setup, context, this::logException, monitor);
   }
 
-  private MavenExecutionPlan calculateExecutionPlan(MavenProject mavenProject, Collection<String> tasks, boolean setup,
+  private static MavenExecutionPlan calculateExecutionPlan(MavenProject mavenProject, Collection<String> tasks,
+      boolean setup, IMavenExecutionContext context, Consumer<? super CoreException> exceptionHandler,
       IProgressMonitor monitor) {
     try {
-      return createExecutionContext().execute(mavenProject, (ctx, mon) -> {
+      return context.execute(mavenProject, (ctx, mon) -> {
         IMavenToolbox toolbox = IMavenToolbox.of(ctx);
         return toolbox.calculateExecutionPlan(tasks, setup);
       }, monitor);
     } catch(CoreException e) {
-      manager.getMarkerManager().addErrorMarkers(pom, IMavenConstants.MARKER_POM_LOADING_ID, e);
+      exceptionHandler.accept(e);
```

Comment:
This part looks suspicious (also in the original) why not just throw the exception and let it handle at the caller site?

Reply:
This method is called by `IMavenProjectFacade.calculateExecutionPlan/setupExecutionPlan()`, which don't permit CoreExceptions in the method signature.
So without breaking the API we cannot propagate the original exception. It would only possible when wrapping it in a RuntimeException.
But I think that would not help much here. It should have been thrown before, but changing it now is not ideal.
But we should note that for a M2E 3.0 :)

Reply:
then wrapping it in a runtime exception (or "handle") it there, this is simply bad style and we should avoid that so pushing the handling as far as possible to the top.